src/metabase/driver/postgres.clj

```diff
@@ -281,6 +287,25 @@
   (and (str/starts-with? database-type "\"")
        (str/ends-with? database-type "\"")))
 
+(defmethod sql.qp/->honeysql [:postgres :convert-timezone]
+  [driver [_ arg target-timezone source-timezone]]
+  (let [expr         (sql.qp/->honeysql driver (cond-> arg
+                                                 (string? arg) u.date/parse))
+        timestamptz? (hx/is-of-type? expr "timestamptz")]
+    (when (and timestamptz? source-timezone)
+      (throw (ex-info (tru "`timestamp with time zone` columns shouldn''t have a `source timezone`")
+                      {:type            qp.error-type/invalid-query
+                       :target-timezone target-timezone
+                       :source-timezone source-timezone})))
+    (let [source-timezone (or source-timezone (qp.timezone/results-timezone-id))
+          expr            (cond->> expr
+                            (and (not timestamptz?) source-timezone)
+                            (hsql/call :timezone source-timezone)
+                            target-timezone
+                            (hsql/call :timezone target-timezone))]
+      (hx/with-type-info expr
+        {::hx/database-type "timestamp"}))))
```

Comment:
`::hx/database-type` shouldn't be used outside the hx namespace. Use `hx/with-database-type-info`.